File: stuff_presentation/src/main/java/android/montreal/devfest/stuff/presentation/presenters/MainPresenterImpl.java
package android.montreal.devfest.stuff.presentation.presenters;

import android.montreal.devfest.stuff.presentation.navigation.Navigator;
import android.montreal.devfest.stuff.usecases.GetRandomStuffException;
import android.montreal.devfest.stuff.usecases.ManageRandomStuffUseCase;
import android.montreal.devfest.stuff.usecases.RandomStuffData;
import android.montreal.devfest.stuff.usecases.SetRandomStuffException;

class MainPresenterImpl implements MainPresenter {
    private final MainPresenter.MainView view;
    private final ManageRandomStuffUseCase manageRandomStuff;
    private Navigator.ViewMore viewMore;

    public MainPresenterImpl(
            MainView view,
            ManageRandomStuffUseCase manageRandomStuff,
            Navigator.ViewMore viewMore
    ) {
        this.view = view;
        this.manageRandomStuff = manageRandomStuff;
        this.viewMore = viewMore;
    }

    @Override
    public void onViewReady() {
        try {
            final RandomStuffData data = this.manageRandomStuff.getRandomStuff();
            this.view.showRandomStuff(data.text, data.number);
        } catch (GetRandomStuffException e) {
            this.view.showMissingError();
        }
    }

    @Override
    public void updateRandomStuff(String text, int value) {
        try {
            this.manageRandomStuff.setRandomStuff(new RandomStuffData(text, value));
            this.view.showUpdateSuccess();
        } catch (SetRandomStuffException e) {
            this.view.showUpdateError();
        }
    }

    @Override
    public void openViewMoreLink() {
        viewMore.showViewMore();
    }
}

File: stuff_presentation/src/main/java/android/montreal/devfest/stuff/presentation/presenters/MainPresenterImpl.kt
package android.montreal.devfest.stuff.presentation.presenters

import android.montreal.devfest.stuff.usecases.GetRandomStuffException
import android.montreal.devfest.stuff.usecases.ManageRandomStuffUseCase
import android.montreal.devfest.stuff.usecases.RandomStuffData
import android.montreal.devfest.stuff.usecases.SetRandomStuffException

internal class MainPresenterImpl(
        private val view: MainPresenter.MainView,
        private val manageRandomStuff: ManageRandomStuffUseCase
) : MainPresenter {

    override fun onViewReady() {
        try {
            val (text, number) = this.manageRandomStuff.getRandomStuff()
            this.view.showRandomStuff(text!!, number)
        } catch (e: GetRandomStuffException) {
            this.view.showMissingError()
        }

    }

    override fun updateRandomStuff(text: String, number: Int) {
        try {
            this.manageRandomStuff.setRandomStuff(RandomStuffData(text, number))
            this.view.showUpdateSuccess()
        } catch (e: SetRandomStuffException) {
            this.view.showUpdateError()
        }
    }
}
